lib/logstash/filters/kv.rb

```diff
@@ -238,27 +239,39 @@ def filter(event)
     # Add default key-values for missing keys
     kv = @default_keys.merge(kv)
 
-    # If we have any keys, create/append the hash
-    if kv.length > 0
-      if @target.nil?
-        # Default is to write to the root of the event.
-        dest = event.to_hash
-      else
-        if !event[@target].is_a?(Hash)
-          @logger.debug("Overwriting existing target field", :target => @target)
-          dest = event[@target] = {}
-        else
-          dest = event[@target]
-        end
+    unless kv.empty?
+      if @target
+        event.remove("#{@target}")
+        @logger.debug? && @logger.debug("Overwriting existing target field", :target => @target)
       end
 
-      dest.merge!(kv)
+      inject_fields(event, @target, kv)
+
       filter_matched(event)
     end
   end
 
   private
 
+  # recursiverly inject the data Hash content into event. the data parameter must be Hash on initial call.
+  #
+  # this is important to do here because we cannot assume in the future the in-place mutability of the Event#to_hash structure.
+  # TODO: (colin) we can certainly optimize this by introducing a new Event API to directly set fields without going through the
```

Comment:
This is essentially [LogStash::Util.hash_merge](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/util.rb#L67) so I'm just trying to avoid creating 2 identical functions.

Also, this function doesn't need to be recursive since it only exists because of the lack of root level merging. Once it unrolls the first level of the hash, it can use normal setter:

```ruby
  def inject_fields(event, path, data)
    case data
    when Hash
      data.each{|k, v| event["#{path}[#{k}]"], v)}
    else
      event[path] = data
    end
  end
```